File: computeAstar.m
function shortcut_course = computeAstar(map, start, goal, size_x, size_y)
    % % -----------A star開始------------ %
    x = start(1, 1);
    y = start(2, 1);
%     pre_x = x;
%     pre_y = y;

%     cost_table = [1, 1, 1;
%                   1, 0, 1;
%                   0, 0, 0];
    cost_table = [0, 0, 0;
                  0, 0, 0;
                  0, 0, 0];

%     open_list = [];
    init_node.x = -1;
    init_node.y = -1;
    init_node.score = uint16(9999);
    init_node.g_cost = uint16(9999);
    init_node.h_cost = uint16(9999);
    open_list = repmat(init_node, 1, 999);
    [map, open_list] = openAroundNodeDP(map, open_list, x, y, cost_table, size_x, size_y, goal); % スタートノードの周りをオープン

    count = 0;
%     for i = 1 : 10000 
    while x ~= goal(1, 1) || y ~= goal(2, 1)

        [x, y] = searchRefNode(open_list); % スコアが最も小さいノードのx, yを得る
%         cost_table = getCostTable(x, y, pre_x, pre_y); % コストテーブルを更新 
        [map, open_list] = openAroundNodeDP(map, open_list, x, y, cost_table, size_x, size_y, goal);

%         pre_x = x;
%         pre_y = y;

        count = count + 1;

    end

    disp(count)

    % % -----------最短経路の座標を取得------------ %
%     map.shorter_path_grid(y, x) = 2;
    store_x = zeros(1, 9999);
    store_y = zeros(1, 9999);
    cnt = 1;
    while x ~= start(1, 1) || y ~= start(2, 1)

        temp_xy = map(y, x).parent;
        x = double(temp_xy(1));
        y = double(temp_xy(2));

        store_x(cnt) = x;
        store_y(cnt) = y;
        cnt = cnt + 1;

%         map.shorter_path_grid(y, x) = 2;
    end

    shortcut_course = [store_x; store_y];
end

File: codegen/lib/computeAstar/computeAstar.c
#include "computeAstar.h"
#include "computeAstar_internal_types.h"
#include "computeAstar_types.h"
#include "openAroundNodeDP.h"
#include <string.h>

// SPDX-License-Identifier: GPL-3.0-only
/*
     Copyright (C) 2021 Haruki Shimotor. All right reserved.
*/
void computeAstar(emxArray_struct0_T *map, const double start[2],
                  const double goal[2], double size_x, double size_y,
                  double shortcut_course[19998])
{
  struct_T arg_data[999];
  struct_T b_arg_data[999];
  struct_T open_list[999];
  double x;
  double y;
  int cnt;
  int temp_xy_idx_0_tmp;
  int temp_xy_idx_0_tmp_tmp;
  unsigned short store_x[9999];
  unsigned short store_y[9999];
  short ii_data[999];
  unsigned short scores[999];
  unsigned short temp_xy_idx_1;
  bool exitg1;
  x = start[0];
  y = start[1];
  for (temp_xy_idx_0_tmp = 0; temp_xy_idx_0_tmp < 999; temp_xy_idx_0_tmp++) {
    open_list[temp_xy_idx_0_tmp].x = -1.0;
    open_list[temp_xy_idx_0_tmp].y = -1.0;
    open_list[temp_xy_idx_0_tmp].score = 9999U;
    open_list[temp_xy_idx_0_tmp].g_cost = 9999U;
    open_list[temp_xy_idx_0_tmp].h_cost = 9999U;
  }
  openAroundNodeDP(map, open_list, start[0], start[1], size_x, size_y, goal);
  while ((x != goal[0]) || (y != goal[1])) {
    for (cnt = 0; cnt < 999; cnt++) {
      scores[cnt] = open_list[cnt].score;
    }
    temp_xy_idx_0_tmp_tmp = scores[0];
    for (cnt = 0; cnt < 998; cnt++) {
      temp_xy_idx_0_tmp = scores[cnt + 1];
      if (temp_xy_idx_0_tmp_tmp > temp_xy_idx_0_tmp) {
        temp_xy_idx_0_tmp_tmp = temp_xy_idx_0_tmp;
      }
    }
    cnt = 0;
    temp_xy_idx_0_tmp = 0;
    exitg1 = false;
    while ((!exitg1) && (temp_xy_idx_0_tmp < 999)) {
      if (scores[temp_xy_idx_0_tmp] == temp_xy_idx_0_tmp_tmp) {
        cnt++;
        ii_data[cnt - 1] = (short)(temp_xy_idx_0_tmp + 1);
        if (cnt >= 999) {
          exitg1 = true;
        } else {
          temp_xy_idx_0_tmp++;
        }
      } else {
        temp_xy_idx_0_tmp++;
      }
    }
    if (1 > cnt) {
      cnt = 0;
    }
    for (temp_xy_idx_0_tmp = 0; temp_xy_idx_0_tmp < cnt; temp_xy_idx_0_tmp++) {
      arg_data[temp_xy_idx_0_tmp] = open_list[ii_data[temp_xy_idx_0_tmp] - 1];
    }
    for (temp_xy_idx_0_tmp = 0; temp_xy_idx_0_tmp < cnt; temp_xy_idx_0_tmp++) {
      b_arg_data[temp_xy_idx_0_tmp] = open_list[ii_data[temp_xy_idx_0_tmp] - 1];
    }
    x = arg_data[0].x;
    y = b_arg_data[0].y;
    openAroundNodeDP(map, open_list, arg_data[0].x, b_arg_data[0].y, size_x,
                     size_y, goal);
  }
  memset(&store_x[0], 0, 9999U * sizeof(unsigned short));
  memset(&store_y[0], 0, 9999U * sizeof(unsigned short));
  cnt = 0;
  while ((x != start[0]) || (y != start[1])) {
    temp_xy_idx_0_tmp = (int)y - 1;
    temp_xy_idx_0_tmp_tmp = (int)x - 1;
    temp_xy_idx_1 =
        map->data[((int)y + map->size[0] * ((int)x - 1)) - 1].parent[1];
    x = map->data[((int)y + map->size[0] * ((int)x - 1)) - 1].parent[0];
    y = temp_xy_idx_1;
    store_x[cnt] =
        map->data[temp_xy_idx_0_tmp + map->size[0] * temp_xy_idx_0_tmp_tmp]
            .parent[0];
    store_y[cnt] = temp_xy_idx_1;
    cnt++;
  }
  for (temp_xy_idx_0_tmp = 0; temp_xy_idx_0_tmp < 9999; temp_xy_idx_0_tmp++) {
    cnt = temp_xy_idx_0_tmp << 1;
    shortcut_course[cnt] = store_x[temp_xy_idx_0_tmp];
    shortcut_course[cnt + 1] = store_y[temp_xy_idx_0_tmp];
  }
}
